### ISSUE TITLE
Trying to flatten an array with unknown depth causes a type error

### ISSUE THREAD

**Issue by GuiltyDolphin** [microsoft/TypeScript#44408]
# Bug Report

Trying to flatten an array with unknown depth causes a type error.

### 🔎 Search Terms

 * `flat`
 * `Infinity`

### 🕗 Version & Regression Information

- Errors in version 4.2.4
- This changed between versions 3.6.2 and 3.7.5 (but probably just because the type definitions aren't supported prior)

The PR https://github.com/microsoft/TypeScript/pull/32131 provides a definition for `FlatArray` that has this issue.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=99&ts=4.3.2&filetype=ts#code/PTAEDMEsCcGcBdQBMCmUB2LQENSYSkqADaQKgDuk8AFjtAEbXTbQCeyKADrQFDxsuWAHIoCSADwAVAHygAvKAAUU0AB9Qo8dJkBKANoBdANy9eAYwD26cq2gAuTWPiEJ6AK4BbBimhzF+gCMADT6AEyhAMyGMaa8ICSW5tjExGzBoLRkoEyI4NiQxLAQltCgnlhKAG6+sJDWoAAsAHRhzY268cBdlNR02OigvtCloADkAETQKObu0BNjoNnT4L4o6OaEyDAz8Gmgo5DoSDvmexxHS-DFlhSDAkI46OiW8Njw9ejNFtbk07CODzeXxGBT0aDNcDEd5KACS6Aw1DYujiCTYlncoGSgyQlhwoAAju5IOYANaUUqk1gY444YrgSypW4An42RD-MKArw+aCgxRKOx0p5sIy6SHQ+BwhFHJEoswJBjuPKjWCWCqgabYVWDBq0LBSQQoADK5mgkB4oC40LYAHMRu5aVlirixOgxnkCsQMorEOjMdihp53BKsNQegwOO46ugbZkaFgGUyqDHOIiPrqaO8lsUUETIFUUutEPA8XqDphLSMqpBUERLmWEAMkKwiKQGCx2PYejR4PAuACQDa+u4GM0rJ5gJ4SSNVeB4MADUITWaeMAGMRLAxgEgABwoACc4AAbPuAKyBfcX7CRRoAdgYDECp9PDH3kR32FPAAYv0fwPu0GwB9b1vHdgDbcDIFHMQwi-C9mjsbA2GaJBmmuABiAAZMJAn4Q1QAAMQlABBaAWDYMIJFI6AMgAEW4WghgADxcY5iiBHl-FAABvXhQH40AJlxTAJkcajgj4gSphmOZRNAajmNYpBigAJRQbBhLSaikIkI5VjKeFMGgaiZEkgSBIAfkIkiyKQyjDN8cTQH0ABaEJQC-DJ3IiUBIgyRoMlPDIjwyW8Mh3DJ9y8zzQECdzAh8wI-NigLYqC2KQtisLYoi2KotAWDDH0eieBoQxTPM8yxLI3gAF9ioYugUBY9ZlNANzQCsoTrBQCZQEcaTZnmEx5TAZIozEfEHiwYZSkcSZpiGhZsw1NA1g2LYTkW84DjKI4tt2fZLmoG47kyfCBheN50y+PDHikZwiPeHQwSe+BtIoiQtBcSRZAyabLHAUB4TTNgZCAA)

### 💻 Code

```ts
// first define a nested list with arbitrary depth
type Nested<T> = (T | Nested<T>)[];

const arr: Nested<number> = [1,[2,[3]]];

// locally, this bit fails for me (version 4.2.4)
//
// with an error '"recur"' is referenced directly or indirectly in its own type annotation.
const res: number[] = arr.flat(Infinity);

// you can do a quick workaround as follows:
const res2: number[] = (arr as any[]).flat(Infinity);

// but for some reason on the TypeScript playground this doesn't fail, but you can emulate it
// by using the following definition that is equivalent to the one provided in the standard library:
// https://github.com/microsoft/TypeScript/blob/d8e9f6951919a347bb155b938a5006f9efabb778/lib/lib.es2019.array.d.ts#L21
type FlatArray2<Arr, Depth extends number> = {
    "done": Arr,
    "recur": Arr extends ReadonlyArray<infer InnerArr>
        ? FlatArray2<InnerArr, [-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][Depth]>
        : Arr
}[Depth extends -1 ? "done" : "recur"];

// causes a type error: '"recur"' is referenced directly or indirectly in its own type annotation.
type TestFlat<T> = FlatArray2<Nested<T>, typeof Infinity>
```

### 🙁 Actual behavior

The code errors with `'"recur"' is referenced directly or indirectly in its own type annotation.`

### 🙂 Expected behavior

The code to pass typechecking.

**Comment by jcalz** [comment]
Now that we have genuine [recursive conditional types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-1.html#recursive-conditional-types), I wonder if this shouldn't be refactored to use them.  Either way, recursion limits exist and problems will happen; I guess someone could hand-tune the definition to try to avoid them:

```ts
type FlatArrayʹ<A, D extends number> =
    D extends -1 ? A :
    A extends ReadonlyArray<infer I> ? FlatArrayʹ<
        I, [-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, ...20[]][D]
    > : A;

interface Array<T> {
    flatʹ<A, D extends number = 1>(
        this: A,
        depth?: D
    ): FlatArrayʹ<A, D>[];
}

const arr: Nested<number> = [1, [2, [3]]];
// I'm not sure what type we expect to see coming out of `arr.flat(Infinity)
// Infinity is just some unknown number as far as the compiler is concerned
// we just don't want a crash 
const res = arr.flatʹ(Infinity); // const res: (number | Nested<number>)[] 🤷‍♂️
```

But I feel like recursion in the standard lib always is going to be prone to this no matter what.  

[Playground link](https://www.typescriptlang.org/play?target=99&ts=4.3.2&filetype=ts#code/C4TwDgpgBAYgNgQ2AQQE6oSQnUAB5kBooARKCAD2AgDsATAZyioFcBbAIwlQD4oBeAKChDipCtXpQAtAEYoAfijIoALkHCl5SrQYAlCAhoB7KnBBoMIHAEsqAM05QAkjwXwk5zLjXDhjwgG0ZQgAGQmlCACZCAGZCABZCAFZCADZCAHZCAA5CAE4w0KhpcKKootiihKLkorSizKKcovyoCMKAOk62-wBdHv8iHu8oHmVFAG5+fhtKVFsEAGNoD0sAFR4Ab2HbRGBcAhFNcQZmdgdeIq4ACmGfYAALKzoxglvhGggwB7kxomGASjGbhQ6E8eEIRC4vUmAF9plRZvMllA9AZjKYVjh1lAtj4dkh9hDRFoJKcOKg+Jcbj47o9nop8G8hB8vvcfsQAWNUPojCYQLBditCcQoT1YfwVu0wKhDMBZeAIO18XtKZLpbL5ZAlbsoAgGAgqCBJlNQJAoAA5CB0Sg0LE8C5XVZQAA+FqtNrt-2h-AWxmtuvQY0t1ogtrJnHtUH8JX8ZX80T6Yv4AHpk047DYrKAoE8oAArJj+uiGFjQJhUADWVEMAHcqIxWOSfX7gFBuQwLgh0NqCVdHBmqFmQP9xlBU1BfVR-e2xldwxTXcGPfOuF6elBAHwbgHZdwCwBIAhMkA8H9TZtT1vcgDmlOjATjCb6uoYk+tkwvPb2V2kI7Hae5aL5UA-F1WiA39eVMKMEzXfhX2VLArjiL9x1A9F+UA10IiA6IoJg3Y4OkYJgkQtM2x5FCANkdDMKg8d+1sTNs1zAsixLMtK2rOsGzOVBoIgc83zg2j6OHb8SL-cC0OA11ohA0j-3jddXWQuSyiU8D5J6ai019FgwCsOAHCMK0qAAclbIwoBYJB7h4vjYI-YIoAAKiKQiRNU1CKMkqBpMU2S1ITGSxP5WNCHciC+igoA)

**Comment by GuiltyDolphin** [comment]
Yeah, it's hard without dependent types. We might be able to get away with certain type operations on numbers (e.g., see work over at https://github.com/granule-project/granule), so that we could capture more information on how we are flattening without restricting the number of levels. With support for natural numbers, we might be able to make use of (semi-dependent?) elimination to get some nice recursion without hitting any limits.

I think for `arr.flat(Infinity)` we would just expect a flattened array type out (for any nested array `Nested<T>`, we'd expect `flat(Infinity)` to try and flatten the array completely, and yield an array `T[]`). If `T` is itself an array, this could lead to issues, but this can be guarded against (e.g., see https://github.com/GuiltyDolphin/functionality/blob/19f4b0be86b71da42bcaed6933ef69e4d9b64a56/src/array.ts#L146).

**Comment by jcalz** [comment]
There's no `Infinity` at the type level, though (#32277), so for now this must be the same as just `number`.
